### DISCUSSION TITLE
Commit history

### DISCUSSION THREAD

**Discussion by jaydeep0303** [community/community#52023] [New to GitHub]
We are a small org using GitHub free account. When i use GitHub Rest API to create a report on commit details for a given time period, i don't get the number of additions, deletions and commit size. 
1) If we move to GitHub Team, would i be able to see these data points?
2) Also, would out of box reports be available that would help me get commit history
3) I am assuming that the commit history would be available for all branches. Could you please confirm.

**Comment by SMelidoni** [comment]
Hi there @jaydeep0303 👋🏽

1. **Access to data points**: Upgrading to a GitHub Team plan will not inherently provide additional data points in the REST API response. To obtain the number of additions, deletions, and commit size, you can use the [compare two commits API](https://docs.github.com/en/rest/reference/repos#compare-two-commits). The response will include details about additions, deletions, and the total number of changed files between the two commits.

2. **Out-of-the-box reports**: GitHub does not offer out-of-the-box reports specifically for commit history. However, you can generate your own reports by utilizing the GitHub REST API or GraphQL API to fetch the necessary data and then process it according to your needs.

3. **Commit history for all branches**: Yes, you can obtain commit history for all branches by using the [list commits API](https://docs.github.com/en/rest/reference/repos#list-commits) and specifying the `sha` parameter to target a specific branch. You can iterate through all branches by first using the [list branches API](https://docs.github.com/en/rest/reference/repos#list-branches) and then fetching the commit history for each branch separately.

Hope this helps!

**Reply by jaydeep0303** [reply]
Thank you @SMelidoni ! Trying it out now!

**Reply by jaydeep0303** [reply]
When i create a fine-grained access token to access my organization info through APIs, i see that a request to list all repos in the org only shows the public repos. But i have provided read access to all repos in my org when creating the fine-grained access token. Any idea why this is happening. I am a owner of the org, so not sure why this would be the behavior.

**Reply by SMelidoni** [reply]
Hi @jaydeep0303,

It seems like you might be encountering an issue with the access token or the API endpoint you are using. Here are a few suggestions to help you troubleshoot the issue:

1. Double-check the access token permissions: Ensure that your access token has the necessary permissions to access the repositories within your organisation. The token should have the `repo` scope enabled, which grants full control of private repositories.

2. Verify the API endpoint: Make sure you're using the correct API endpoint to list all repositories in your organisation. To list all repositories within an organisation, you should use the following endpoint:

`GET /orgs/:org/repos`

Replace `:org` with your organisations name. Remember to add the access token to your request using the `Authorization` header:

`Authorization: token your_access_token`

3. Pagination: If your organisation has a large number of repositories, the API response might be paginated. In this case, you'll need to check the `Link` header in the response for the next page of results. You can also increase the number of results per page by using the `per_page` query parameter (up to a maximum of 100).

If you're still having issues, please share the exact API request you're making (without the access token), and any relevant error messages or response details. This will help in identifying the root cause of the problem and providing a suitable solution.

I hope this helps! If you have any further questions or need assistance, please let me know.

**Reply by jaydeep0303** [reply]
Thank you once again @SMelidoni. My token had expired. Am able to hit the repos now after generating a new one!

**Reply by jaydeep0303** [reply]
Now i have the following snippet to get commits: 
```
    for branch in branches:
        # Get commits for the branch in the given time period
        since_date = "2022-01-01T00:00:00Z"
        until_date = "2022-01-31T23:59:59Z"
        try:
            commits = repo.get_commits(since=since_date, until=until_date, sha=branch.commit.sha)
        except Exception as e:
            print(f"Error getting commits for branch {branch.name}: {e}")
            continue
```
but it is not returning any commits. I can see that there are commits in my repos during this time period. I have checked that i am able to access the repo and its branches. Any suggestions please?

**Reply by SMelidoni** [reply]
Hi @jaydeep0303,

It seems like you're using the PyGithub library to interact with the GitHub API. The code snippet you provided looks correct for fetching commits within a specific date range for each branch. Here are a few suggestions to help you debug the issue:

1. Print debugging information: Add print statements to your code to display the branches and commits you're iterating through. This will help you identify if the issue is related to accessing the branches or fetching the commits.

```python
for branch in branches:
    print(f"Processing branch: {branch.name}")
    since_date = "2022-01-01T00:00:00Z"
    until_date = "2022-01-31T23:59:59Z"
    try:
        commits = repo.get_commits(since=since_date, until=until_date, sha=branch.commit.sha)
        print(f"Commits fetched: {commits.totalCount}")
    except Exception as e:
        print(f"Error getting commits for branch {branch.name}: {e}")
        continue
```

2. Test the API call directly: You can test the API call directly by using a tool like `curl` or Postman to send a request to the GitHub API. This will help you confirm if the issue is related to the API call itself or the PyGithub library. Here's an example `curl` command to fetch commits for a specific branch within a specific date range:

```
curl -H "Authorization: token your_access_token" \
  "https://api.github.com/repos/your-username/your-repo/commits?sha=branch_name&since=2022-01-01T00:00:00Z&until=2022-01-31T23:59:59Z"
```

Replace `your_access_token`, `your-username`, `your-repo`, and `branch_name` with the appropriate values.

3. Check the date format: Ensure that the date format you're using in the `since` and `until` parameters is correct. The format should be ISO 8601, as you have specified in your example. However, double-check to make sure there are no typos or formatting errors.
If you're still having issues, please provide any additional debugging information or error messages you've encountered. This will help in identifying the root cause of the problem and providing a suitable solution.

I hope this helps! If you have any further questions please let me know.

**Reply by jaydeep0303** [reply]
Hi, just got back to this issue. Firstly, thank you so much for the help till now. 
When i use the API GET /repos/owner/repo/stats/contributors, i get only the number of commits per contributor, but the additions and deletions is always showing as 0. Could you please throw some light on why this would be the case?

**Reply by jaydeep0303** [reply]
Adding another point.  I managed to solve the problem by listing commits and then getting stats on each commit. This provides me with addition and deletion. 
However, it is providing these stats only for main branch. My requirement is to get commit stats on feature branches as well. How can i make that work?

**Reply by SMelidoni** [reply]
Hi @jaydeep0303,

The `/repos/owner/repo/stats/contributors` endpoint provides a summary of each contributor's activity, and it is possible that it might not return the additions and deletions in some cases. The data returned by this endpoint might also be cached and not reflect the latest changes.

To get commit stats on feature branches, you can use the `GET /repos/owner/repo/commits` endpoint and pass the `sha` parameter, which specifies the branch name you want to list commits for. Here's an example:

`GET /repos/owner/repo/commits?sha=feature_branch`

This will hopefully list commits for the `feature_branch` branch. You can then iterate through each commit to get the stats as you have done for the main branch.

I hope this helps! Let me know if you have any further questions.
